File: src-tools/lib/command/configure.js
'use strict';

const { configureSources } = require('../configure/configure_sources');
const { pathJoin, readFileUtf8, readFileYaml } = require('../util/fs');
const { parse: parseYaml } = require('../util/yaml');
const { stripLastNewline, normalizeNewlines } = require('../util/string_util');
const { createBareObject } = require('../util/bare');
const { createTempDir } = require('../util/fs');
const { assert } = require('../util/assert');

const configureCommandSpec = createBareObject({
    description: 'Prepare Duktape source and header files for application build',
    optionsInitCallback: (opts) => {
        opts.forcedOptions = createBareObject({});
        opts.fixupLines = [];
        opts.userBuiltinFiles = [];
    },
    options: createBareObject({
        'source-directory': { type: 'path', default: void 0, description: 'Directory with raw input sources (defaulted based on script path)' },
        'output-directory': { type: 'path', default: void 0, required: true, description: 'Directory for output files (created automatically, must not exist)'},
        'temp-directory': { type: 'path', default: void 0, description: 'Force a specific temp directory (default is to create automatically)' },
        'config-directory': { type: 'path', default: void 0, description: 'Directory with config metadata (defaulted based on script path)' },
        'config-metadata': { type: 'path', default: void 0, deprecated: true, description: 'Deprecated version of --config-directory, do not use' },
        'option-file': { type: 'path', repeat: true, callback: (v, opts) => {
            let doc = readFileYaml(v);
            Object.assign(opts.forcedOptions, doc);
        }, description: 'YAML file(s) providing config option overrides (may be given multiple times)' },
        'option-yaml': { type: 'string', repeat: true, callback: (v, opts) => {
            let doc = parseYaml(v);
            Object.assign(opts.forcedOptions, doc);
        }, description: 'Override option(s) using inline YAML, e.g. --option-yaml "DUK_USE_DEEP_C_STACK: true"' },
        'define': { type: 'string', short: 'D', repeat: true, callback: (v, opts) => {
            let m;
            // -DFOO(a,b,c)=((a)+(b)+(c))
            m = /^([A-Z0-9_]+)\((.*?)\)=(.*?)$/.exec(v);
            if (m) {
                let defName = m[1];
                let defArgs = m[2];
                let defValue = m[3];
                opts.forcedOptions[defName] = { verbatim: '#define ' + defName + '(' + defArgs + ') ' + defValue };
                return;
            }
            // -DFOO=123
            // -DFOO="foo"
            m = /^([A-Z0-9_]+)=(.*?)$/.exec(v);
            if (m) {
                let defName = m[1];
                let defValue = m[2];
                opts.forcedOptions[defName] = defValue;
                return;
            }
            // -DFOO
            m = /^([A-Z0-9_]+)$/.exec(v);
            if (m) {
                let defName = m[1];
                opts.forcedOptions[defName] = true;
                return;
            }
            throw new TypeError('invalid --define option value: ' + v);
        }, description: 'Override option using a C compiler like syntax, e.g. "--define FOO", "-DFOO=123", or "-DFOO(a,b)=((a)+(b))"' },
        'undefine': { type: 'string', short: 'U', repeat: true, callback: (v, opts) => {
            // -UFOO
            // No other variants for undefining.
            let m;
            m = /^([A-Z0-9_]+)$/.exec(v);
            if (m) {
                let defName = m[1];
                opts.forcedOptions[defName] = false;
                return;
            }
            throw new TypeError('invalid --undefine option value: ' + v);
        }, description: 'Override option using a C compiler like syntax, e.g. "--undefine FOO" or "-UFOO"' },
        'fixup-line': { type: 'string', repeat: true, callback: (v, opts) => {
            opts.fixupLines.push(v);
        }, description: 'C header fixup line to be appended to generated header, e.g. --fixup-line "#define DUK_USE_FASTINT"' },
        'fixup-file': { type: 'path', repeat: true, callback: (v, opts) => {
            let data = stripLastNewline(normalizeNewlines(readFileUtf8(v)));
            data.split('\n').forEach((line) => {
                opts.fixupLines.push(line);
            });
        }, description: 'C header snippet file to be appended to generated header, useful for manual option fixups, e.g. --fixup-file my_fixups.h' },
        'builtin-file': { type: 'path', repeat: true, callback: (v, opts) => {
            opts.userBuiltinFiles.push(v);
        }, description: 'Built-in string/object YAML metadata to be applied over default built-ins (multiple files may be given, applied in sequence)' },
        'separate-sources': { type: 'boolean', default: false, value: true, deprecated: true, description: 'Output separate sources instead of amalgamated source (default is amalgamated)' },
        'line-directives': { type: 'boolean', default: false, value: true, description: 'Output #line directives in amalgamated source (default is false)' },
        'dll': { type: 'boolean', default: false, value: true, description: 'Enable DLL build of Duktape, affects symbol visibility macros especially on Windows' },
        'c99-types-only': { type: 'boolean', default: false, value: true, description: 'Assume C99 types, no legacy type detection' },
        'use-cpp-warning': { type: 'boolean', default: false, value: true, description: 'Emit a (non-portable) #warning when appropriate, e.g. for config setting mismatches' },
        'rom-support': { type: 'boolean', default: false, value: true, description: 'Add support for ROM strings/objects' },
        'rom-auto-lightfunc': { type: 'boolean', default: false, value: true, description: 'Convert ROM built-in function properties into lightfuncs automatically whenever possible' },
        'sanity-warning': { type: 'boolean', default: false, value: true, description: 'Emit a warning instead of #error for option sanity check issues (combine with --use-cpp-warning)' },
        'omit-removed-config-options': { type: 'boolean', default: false, value: true, description: 'Omit removed config options from generated headers' },
        'omit-deprecated-config-options': { type: 'boolean', default: false, value: true, description: 'Omit deprecated config options from generated headers' },
        'omit-unused-config-options': { type: 'boolean', default: false, value: true, description: 'Omit unused config options from generated headers' },
        'emit-config-sanity-check': { type: 'boolean', default: false, value: true, description: 'Emit preprocessor checks for config option consistency (DUK_USE_xxx)' },
        'emit-legacy-feature-check': { type: 'boolean', default: false, value: true, deprecated: true, description: 'Emit preprocessor checks to reject legacy feature options (DUK_OPT_xxx)' }
    })
});
exports.configureCommandSpec = configureCommandSpec;

function configureCommand(cmdline, autoDuktapeRoot) {
    var opts = cmdline.commandOpts;

    if (!autoDuktapeRoot && opts['source-directory']) {
        autoDuktapeRoot = pathJoin(opts['source-directory'], '..');
    }
    var sourceDirectory = opts['source-directory'] || pathJoin(autoDuktapeRoot, 'src-input');
    var outputDirectory = assert(opts['output-directory'], '--output-directory required');
    var configDirectory = opts['config-directory'] || opts['config-metadata'] || pathJoin(autoDuktapeRoot, 'config');
    var tempDirectory = assert(opts['temp-directory'] || createTempDir());
    var licenseFile = opts['license-file'] || pathJoin(autoDuktapeRoot, 'LICENSE.txt');
    var authorsFile= opts['authors-file'] || pathJoin(autoDuktapeRoot, 'AUTHORS.rst');
    var unicodeDataFile = opts['unicode-data'] || pathJoin(sourceDirectory, 'UnicodeData.txt');
    var specialCasingFile = opts['special-casing'] || pathJoin(sourceDirectory, 'SpecialCasing.txt');
    var forcedOptions = assert(opts.forcedOptions);
    var fixupLines = assert(opts.fixupLines);
    var userBuiltinFiles = assert(opts.userBuiltinFiles);

    if (opts['separate-sources']) {
        throw new TypeError('separate sources no longer supported');
    }
    if (opts['emit-legacy-feature-check']) {
        throw new TypeError('--emit-legacy-feature-check no longer supported');
    }
    if (opts['rom-support'] === false) {
        console.log('assuming --rom-support (now always enabled)');
        opts['rom-support'] = true;
    }

    return configureSources({
        sourceDirectory,
        outputDirectory,
        configDirectory,
        tempDirectory,
        licenseFile,
        authorsFile,
        unicodeDataFile,
        specialCasingFile,
        dukDistMetaFile: opts['duk-dist-meta'],
        gitCommit: opts['git-commit'],
        gitDescribe: opts['git-describe'],
        gitBranch: opts['git-branch'],
        platform: opts['platform'],
        architecture: opts['architecture'],
        compiler: opts['compiler'],
        forcedOptions,
        fixupLines,
        userBuiltinFiles,
        romAutoLightFunc: opts['rom-auto-lightfunc'],
        lineDirectives: opts['line-directives'],
        c99TypesOnly: opts['c99-types-only'],
        dll: opts['dll'],
        romSupport: opts['rom-support'],
        emitConfigSanityCheck: opts['emit-config-sanity-check'],
        sanityStrict: !opts['sanity-warning'],
        useCppWarning: opts['use-cpp-warning'],
        omitRemovedConfigOptions: opts['omit-removed-config-options'],
        omitDeprecatedConfigOptions: opts['omit-deprecated-config-options'],
        omitUnusedConfigOptions: opts['omit-unused-config-options']
    });
}
exports.configureCommand = configureCommand;

File: tools/configure.py
#!/usr/bin/env python
#
#  Compatibility stub which now executes JS-based tooling.
#
#  Should be Python2 and Python3 compatible.

import os
import sys
import time
import subprocess
import optparse
import yaml
import tempfile

def detect_nodejs():
    try:
        cmd = [ 'nodejs', '-e', 'console.log("test")' ]
        res = subprocess.check_output(cmd)
        if res[:4] == 'test'.encode('utf-8'):
            return 'nodejs'
    except:
        pass

    try:
        cmd = [ 'node', '-e', 'console.log("test")' ]
        res = subprocess.check_output(cmd)
        if res[:4] == 'test'.encode('utf-8'):
            return 'node'
    except:
        pass

    return None

def main():
    sys.stderr.write('\n')
    sys.stderr.write('****************************************************************************\n')
    sys.stderr.write('*** Duktape python tooling is obsolete, migrate to JS-based tooling!     ***\n')
    sys.stderr.write('*** This tool now internally invokes the JS-based tooling.               ***\n')
    sys.stderr.write('*** Minimum Node.js version is 14.x.                                     ***\n')
    sys.stderr.write('****************************************************************************\n')
    sys.stderr.write('\n')
    time.sleep(2)

    parser = optparse.OptionParser(
        usage='Usage: %prog [options]',
        description='Compatibility stub for JS-based tooling'
    )

    # Forced options from multiple sources are gathered into a shared list
    # so that the override order remains the same as on the command line.
    force_options_yaml = []
    def add_force_option_yaml(option, opt, value, parser):
        force_options_yaml.append(value)
    def add_force_option_file(option, opt, value, parser):
        with open(value, 'rb') as f:
            force_options_yaml.append(f.read())
    def add_force_option_define(option, opt, value, parser):
        defname, eq, defval = value.partition('=')
        if not eq:
            doc = { defname: True }
        else:
            defname, paren, defargs = defname.partition('(')
            if not paren:
                doc = { defname: defval }
            else:
                doc = { defname: { 'verbatim': '#define %s%s%s %s' % (defname, paren, defargs, defval) } }
        force_options_yaml.append(yaml.safe_dump(doc))
    def add_force_option_undefine(option, opt, value, parser):
        tmp = value.split('=')
        if len(tmp) == 1:
            doc = { tmp[0]: False }
        else:
            raise Exception('invalid option value: %r' % value)
        force_options_yaml.append(yaml.safe_dump(doc))

    fixup_header_lines = []
    def add_fixup_header_line(option, opt, value, parser):
        fixup_header_lines.append(value)
    def add_fixup_header_file(option, opt, value, parser):
        with open(value, 'rb') as f:
            for line in f:
                if line[-1] == '\n':
                    line = line[:-1]
                fixup_header_lines.append(line)

    # Log level options.
    parser.add_option('--quiet', dest='quiet', action='store_true', default=False, help='Suppress info messages (show warnings)')
    parser.add_option('--verbose', dest='verbose', action='store_true', default=False, help='Show verbose debug messages')

    # Options for configure.py tool itself.
    parser.add_option('--source-directory', dest='source_directory', default=None, help='Directory with raw input sources (defaulted based on configure.py script path)')
    parser.add_option('--output-directory', dest='output_directory', default=None, help='Directory for output files (created automatically if it doesn\'t exist, reused if safe)')
    parser.add_option('--license-file', dest='license_file', default=None, help='Source for LICENSE.txt (defaulted based on configure.py script path)')
    parser.add_option('--authors-file', dest='authors_file', default=None, help='Source for AUTHORS.rst (defaulted based on configure.py script path)')
    parser.add_option('--git-commit', dest='git_commit', default=None, help='Force git commit hash')
    parser.add_option('--git-describe', dest='git_describe', default=None, help='Force git describe')
    parser.add_option('--git-branch', dest='git_branch', default=None, help='Force git branch name')
    parser.add_option('--duk-dist-meta', dest='duk_dist_meta', default=None, help='duk_dist_meta.json to read git commit etc info from')

    # Options for combining sources.
    parser.add_option('--separate-sources', dest='separate_sources', action='store_true', default=False, help='Output separate sources instead of combined source (default is combined)')
    parser.add_option('--line-directives', dest='line_directives', action='store_true', default=False, help='Output #line directives in combined source (default is false)')

    # Options for built-ins.
    parser.add_option('--rom-support', dest='rom_support', action='store_true', help='Add support for ROM strings/objects (increases duktape.c size considerably)')
    parser.add_option('--rom-auto-lightfunc', dest='rom_auto_lightfunc', action='store_true', default=False, help='Convert ROM built-in function properties into lightfuncs automatically whenever possible')
    parser.add_option('--user-builtin-metadata', dest='obsolete_builtin_metadata', default=None, help=optparse.SUPPRESS_HELP)
    parser.add_option('--builtin-file', dest='builtin_files', metavar='FILENAME', action='append', default=[], help='Built-in string/object YAML metadata to be applied over default built-ins (multiple files may be given, applied in sequence)')

    # Options for Unicode.
    parser.add_option('--unicode-data', dest='unicode_data', default=None, help='Provide custom UnicodeData.txt')
    parser.add_option('--special-casing', dest='special_casing', default=None, help='Provide custom SpecialCasing.txt')

    # Options for genconfig.py.
    parser.add_option('--config-metadata', dest='config_metadata', default=None, help='metadata directory (defaulted based on configure.py script path)')
    parser.add_option('--platform', dest='platform', default=None, help='platform (default is autodetect)')
    parser.add_option('--compiler', dest='compiler', default=None, help='compiler (default is autodetect)')
    parser.add_option('--architecture', dest='architecture', default=None, help='architecture (default is autodetec)')
    parser.add_option('--c99-types-only', dest='c99_types_only', action='store_true', default=False, help='assume C99 types, no legacy type detection')
    parser.add_option('--dll', dest='dll', action='store_true', default=False, help='dll build of Duktape, affects symbol visibility macros especially on Windows')
    parser.add_option('--support-feature-options', dest='support_feature_options', action='store_true', default=False, help=optparse.SUPPRESS_HELP)
    parser.add_option('--emit-legacy-feature-check', dest='emit_legacy_feature_check', action='store_true', default=False, help='emit preprocessor checks to reject legacy feature options (DUK_OPT_xxx)')
    parser.add_option('--emit-config-sanity-check', dest='emit_config_sanity_check', action='store_true', default=False, help='emit preprocessor checks for config option consistency (DUK_USE_xxx)')
    parser.add_option('--omit-removed-config-options', dest='omit_removed_config_options', action='store_true', default=False, help='omit removed config options from generated headers')
    parser.add_option('--omit-deprecated-config-options', dest='omit_deprecated_config_options', action='store_true', default=False, help='omit deprecated config options from generated headers')
    parser.add_option('--omit-unused-config-options', dest='omit_unused_config_options', action='store_true', default=False, help='omit unused config options from generated headers')
    parser.add_option('--add-active-defines-macro', dest='add_active_defines_macro', action='store_true', default=False, help='add DUK_ACTIVE_DEFINES macro, for development only')
    parser.add_option('--define', type='string', metavar='OPTION', dest='force_options_yaml', action='callback', callback=add_force_option_define, default=force_options_yaml, help='force #define option using a C compiler like syntax, e.g. "--define DUK_USE_DEEP_C_STACK" or "--define DUK_USE_TRACEBACK_DEPTH=10"')
    parser.add_option('-D', type='string', metavar='OPTION', dest='force_options_yaml', action='callback', callback=add_force_option_define, default=force_options_yaml, help='synonym for --define, e.g. "-DDUK_USE_DEEP_C_STACK" or "-DDUK_USE_TRACEBACK_DEPTH=10"')
    parser.add_option('--undefine', type='string', metavar='OPTION', dest='force_options_yaml', action='callback', callback=add_force_option_undefine, default=force_options_yaml, help='force #undef option using a C compiler like syntax, e.g. "--undefine DUK_USE_DEEP_C_STACK"')
    parser.add_option('-U', type='string', metavar='OPTION', dest='force_options_yaml', action='callback', callback=add_force_option_undefine, default=force_options_yaml, help='synonym for --undefine, e.g. "-UDUK_USE_DEEP_C_STACK"')
    parser.add_option('--option-yaml', type='string', metavar='YAML', dest='force_options_yaml', action='callback', callback=add_force_option_yaml, default=force_options_yaml, help='force option(s) using inline YAML (e.g. --option-yaml "DUK_USE_DEEP_C_STACK: true")')
    parser.add_option('--option-file', type='string', metavar='FILENAME', dest='force_options_yaml', action='callback', callback=add_force_option_file, default=force_options_yaml, help='YAML file(s) providing config option overrides')
    parser.add_option('--fixup-file', type='string', metavar='FILENAME', dest='fixup_header_lines', action='callback', callback=add_fixup_header_file, default=fixup_header_lines, help='C header snippet file(s) to be appended to generated header, useful for manual option fixups')
    parser.add_option('--fixup-line', type='string', metavar='LINE', dest='fixup_header_lines', action='callback', callback=add_fixup_header_line, default=fixup_header_lines, help='C header fixup line to be appended to generated header (e.g. --fixup-line "#define DUK_USE_FASTINT")')
    parser.add_option('--sanity-warning', dest='sanity_strict', action='store_false', default=True, help='emit a warning instead of #error for option sanity check issues')
    parser.add_option('--use-cpp-warning', dest='use_cpp_warning', action='store_true', default=False, help='emit a (non-portable) #warning when appropriate')
    parser.add_option('--nodejs-command', dest='nodejs_command', default=None, help='Force Node.js command name')

    entry_cwd = os.getcwd()
    script_path = sys.path[0]  # http://stackoverflow.com/questions/4934806/how-can-i-find-scripts-directory-with-python

    (opts, args) = parser.parse_args()
    if len(args) > 0:
        raise Exception('unexpected arguments: %r' % args)

    if opts.obsolete_builtin_metadata is not None:
        raise Exception('--user-builtin-metadata has been removed, use --builtin-file instead')

    if opts.nodejs_command is None:
        nodejs_command = detect_nodejs()
    else:
        nodejs_command = opts.nodejs_command
    if nodejs_command is None:
        raise Exception('failed to detect Node.js, override with --nodejs-command')

    duktool_path = None
    for fn in [
        os.path.join(script_path, 'duktool.js'),
        os.path.join(script_path, '..', 'src-tools', 'index.js'),
        os.path.join(script_path, '..', 'src-tools', 'duktool.js')
    ]:
        if os.path.isfile(fn):
            duktool_path = fn
            break
    if duktool_path is None:
        raise Exception('could not find duktool.js or src-tools/index.js')

    cmd = [
        nodejs_command,
        duktool_path,
        'configure'
    ]
    if opts.output_directory is not None:
        cmd += [ '--output-directory', opts.output_directory ]
    if opts.source_directory is not None:
        cmd += [ '--source-directory', opts.source_directory ]
    else:
        src_dir = os.path.join(script_path, '..', 'src-input')
        if os.path.isdir(src_dir) and os.path.isfile(os.path.join(src_dir, 'duktape.h.in')):
            cmd += [ '--source-directory', src_dir ]
    if opts.config_metadata is not None:
        cmd += [ '--config-directory', opts.config_metadata ]

    forced = {}
    for i in force_options_yaml:
        doc = yaml.safe_load(i)
        for k,v in doc.items():
            forced[k] = v
    opts_fd, opts_fn = tempfile.mkstemp()
    with os.fdopen(opts_fd, 'wb') as f:
        f.write(yaml.safe_dump(forced).encode('utf-8'))
    cmd += [ '--option-file', opts_fn ]

    fixup_fd, fixup_fn = tempfile.mkstemp()
    with os.fdopen(fixup_fd, 'wb') as f:
        f.write(('\n'.join(fixup_header_lines) + '\n').encode('utf-8'))
    cmd += [ '--fixup-file', fixup_fn ]

    for i in opts.builtin_files:
        cmd += [ '--builtin-file', i ]

    if opts.line_directives:
        cmd += [ '--line-directives' ]

    if opts.platform is not None:
        cmd += [ '--platform', opts.platform ]
    if opts.compiler is not None:
        cmd += [ '--compiler', opts.compiler ]
    if opts.architecture is not None:
        cmd += [ '--architecture', opts.architecture ]

    if opts.dll:
        cmd += [ '--dll' ]
    if opts.c99_types_only:
        cmd += [ '--c99-types-only' ]

    sys.stderr.write('*** Executing JS-based tooling with command: ' + repr(cmd) + '\n\n')
    subprocess.check_call(cmd)

if __name__ == '__main__':
    main()
